File: sdk/typescript/src/server/subscriptions.ts
import { listQuery, type HttpTransport } from "../http.js";
import type { Benefits, ListParams, ProviderOverrides, Subscription, SubscriptionPlan } from "../types.js";

export class ServerSubscriptionsService {
  constructor(private readonly http: HttpTransport) {}

  async createPlan(input: {
    code: string;
    name: string;
    amount: string;
    currency: string;
    interval: string;
    interval_days?: string;
    grace_days?: number;
    trial_days?: number;
    benefits?: Benefits;
    provider_overrides?: ProviderOverrides;
  }): Promise<SubscriptionPlan> {
    return this.http.request<SubscriptionPlan>("POST", "/v1/server/subscriptions/plans", {
      auth: "apiKey",
      body: input,
    });
  }

  async listPlans(params?: ListParams): Promise<SubscriptionPlan[]> {
    const res = await this.http.request<{ plans: SubscriptionPlan[] }>(
      "GET",
      "/v1/server/subscriptions/plans",
      { auth: "apiKey", query: listQuery(params) }
    );
    return res.plans ?? [];
  }

  async getPlan(planId: string): Promise<SubscriptionPlan> {
    return this.http.request<SubscriptionPlan>("GET", `/v1/server/subscriptions/plans/${planId}`, {
      auth: "apiKey",
    });
  }

  async updatePlan(
    planId: string,
    input: {
      name?: string;
      amount?: string;
      currency?: string;
      interval?: string;
      interval_days?: string;
      grace_days?: number;
      trial_days?: number;
      benefits?: Benefits;
      provider_overrides?: ProviderOverrides;
      status?: string;
    }
  ): Promise<SubscriptionPlan> {
    return this.http.request<SubscriptionPlan>("PATCH", `/v1/server/subscriptions/plans/${planId}`, {
      auth: "apiKey",
      body: { plan_id: planId, ...input },
    });
  }

  async deletePlan(planId: string): Promise<void> {
    await this.http.request<void>("DELETE", `/v1/server/subscriptions/plans/${planId}`, {
      auth: "apiKey",
    });
  }

  async listSubscriptions(params?: ListParams): Promise<Subscription[]> {
    const res = await this.http.request<{ subscriptions: Subscription[] }>(
      "GET",
      "/v1/server/subscriptions",
      { auth: "apiKey", query: listQuery(params) }
    );
    return res.subscriptions ?? [];
  }

  async getSubscription(subscriptionId: string): Promise<Subscription> {
    return this.http.request<Subscription>("GET", `/v1/server/subscriptions/${subscriptionId}`, {
      auth: "apiKey",
    });
  }

  async cancelSubscription(subscriptionId: string, reason?: string): Promise<Subscription> {
    return this.http.request<Subscription>(
      "POST",
      `/v1/server/subscriptions/${subscriptionId}:cancel`,
      { auth: "apiKey", body: { subscription_id: subscriptionId, reason } }
    );
  }

  async expireSubscription(subscriptionId: string, reason?: string): Promise<Subscription> {
    return this.http.request<Subscription>(
      "POST",
      `/v1/server/subscriptions/${subscriptionId}:expire`,
      { auth: "apiKey", body: { subscription_id: subscriptionId, reason } }
    );
  }
}

File: sdk/go/server/subscriptions.go
package server

import (
	"context"

	serverv1 "github.com/torchwooddev/torchwood/genproto/server/v1"
	sharedv1 "github.com/torchwooddev/torchwood/genproto/shared/v1"
)

// SubscriptionsService 封装 Server API 的订阅管理。
type SubscriptionsService struct {
	c   *Client
	api serverv1.SubscriptionsServiceClient
}

func (s *SubscriptionsService) CreatePlan(ctx context.Context, req *serverv1.CreatePlanRequest) (*serverv1.SubscriptionPlan, error) {
	return s.api.CreatePlan(ctx, req)
}

func (s *SubscriptionsService) ListPlans(ctx context.Context, req *sharedv1.ListRequest) (*serverv1.ListPlansResponse, error) {
	return s.api.ListPlans(ctx, req)
}

func (s *SubscriptionsService) GetPlan(ctx context.Context, planID string) (*serverv1.SubscriptionPlan, error) {
	return s.api.GetPlan(ctx, &serverv1.GetPlanRequest{PlanId: planID})
}

func (s *SubscriptionsService) UpdatePlan(ctx context.Context, req *serverv1.UpdatePlanRequest) (*serverv1.SubscriptionPlan, error) {
	return s.api.UpdatePlan(ctx, req)
}

func (s *SubscriptionsService) DeletePlan(ctx context.Context, planID string) error {
	_, err := s.api.DeletePlan(ctx, &serverv1.DeletePlanRequest{PlanId: planID})
	return err
}

func (s *SubscriptionsService) ListSubscriptions(ctx context.Context, req *sharedv1.ListRequest) (*serverv1.ListSubscriptionsResponse, error) {
	return s.api.ListSubscriptions(ctx, req)
}

func (s *SubscriptionsService) GetSubscription(ctx context.Context, subscriptionID string) (*serverv1.Subscription, error) {
	return s.api.GetSubscription(ctx, &serverv1.GetSubscriptionRequest{SubscriptionId: subscriptionID})
}

func (s *SubscriptionsService) CancelSubscription(ctx context.Context, req *serverv1.CancelSubscriptionRequest) (*serverv1.Subscription, error) {
	return s.api.CancelSubscription(ctx, req)
}

func (s *SubscriptionsService) ExpireSubscription(ctx context.Context, req *serverv1.ExpireSubscriptionRequest) (*serverv1.Subscription, error) {
	return s.api.ExpireSubscription(ctx, req)
}
